### PR TITLE
Implement import and export functionality

### LINKED ISSUE #41: Settings import/export
Since the mime type associations can be somewhat cumbersome to set up by hand, it would be nice to have a way to save / load them from a file.

For inspiration you could look at uBlock, it has ['back up to file'](https://github.com/gorhill/uBlock/blob/dd49e0ad7cc5766c72c28daf8cbe9f2a156dc267/src/js/settings.js#L107) and ['restore from file'](https://github.com/gorhill/uBlock/blob/dd49e0ad7cc5766c72c28daf8cbe9f2a156dc267/src/js/settings.js#L36) features.

### PR DESCRIPTION
**Pull request by Mathnerd314**
Fixes #41.

### CODE REVIEW

File: extension/options.html [hunk 1]
```diff
@@ -80,6 +86,13 @@ <h1>Open in Browser</h1>
 handler for the given document.</p>
 
 <h2>Preferences</h2>
+<div class="row">
+  <button id="export" type="button">Export</button>
+  <button id="import" type="button">Import</button>
```
Comment (Rob--W) [inline extension/options.html]: These buttons do not just export/import "preferences", but all settings. So please move it to a new section after the last row.

File: extension/options.html [hunk 2]
```diff
@@ -80,6 +86,13 @@ <h1>Open in Browser</h1>
...
+<div class="hidden">
+    <input id="importFilePicker" type="file" accept="text/plain">
+</div>
```
Comment (Rob--W) [inline extension/options.html]: Please indent the last three lines.

File: extension/options.html [hunk 3]
```diff
@@ -80,6 +86,13 @@ <h1>Open in Browser</h1>
 handler for the given document.</p>
 
 <h2>Preferences</h2>
+<div class="row">
+  <button id="export" type="button">Export</button>
+  <button id="import" type="button">Import</button>
+<div class="hidden">
```
Comment (Rob--W) [inline extension/options.html]: Use the `hidden` attribute instead of a custom CSS class.

File: extension/options.js [hunk 4]
```diff
@@ -107,3 +107,35 @@ Prefs.setPrefHandler('mime-mappings', renderMimeMappings);
...
 
+$('export').onclick = function exportPrefs() {
+    var settings = Prefs.exportPrefs();
+    var settingsURL = 'data:text/plain;charset=utf-8,' +
+                   encodeURIComponent(settings);
+
+    var a = document.createElement('a');
+    a.href = settingsURL;
+    a.setAttribute('download', 'oib-prefs.txt');
+    a.dispatchEvent(new MouseEvent('click'));
```
Comment (Rob--W) [inline extension/options.js]: `a.click()`

File: extension/options.js [hunk 5]
```diff
@@ -107,3 +107,35 @@ Prefs.setPrefHandler('mime-mappings', renderMimeMappings);
...
+
+$('import').onclick = function importPrefs() {
+    var input = $('importFilePicker');
+    input.value = '';
+    input.click();
+};
+
+$('importFilePicker').onchange = function reallyImportPrefs() {
+    var file = $('importFilePicker').files[0];
+    if ( file === undefined || file.name === '' ) {
```
Comment (Rob--W) [inline extension/options.js]: `if (!file) {`

File: extension/options.js [hunk 6]
```diff
@@ -107,3 +107,35 @@ Prefs.setPrefHandler('mime-mappings', renderMimeMappings);
...
+        return;
+    }
+    if ( file.type.indexOf('text') !== 0 ) {
```
Comment (Rob--W) [inline extension/options.js]: Remove this check.

File: extension/prefs.js [hunk 7]
```diff
@@ -174,6 +176,21 @@ function removeMimeAction(mimeType, isSniffingMimeType) {
     }
 }
 
+function exportPrefs() {
+    return JSON.stringify(prefs, null, 2);
+}
+function importPrefs(newPrefsString) {
+    var newPrefs = JSON.parse(newPrefsString);
+    if (typeof newPrefs !== 'object') {
+        throw 'Invalid';
```
Comment (Rob--W) [inline extension/prefs.js]: `throw new Error('Invalid format; expected an object in JSON');`

File: extension/prefs.js [hunk 8]
```diff
@@ -174,6 +176,21 @@ function removeMimeAction(mimeType, isSniffingMimeType) {
     }
 }
 
+function exportPrefs() {
+    return JSON.stringify(prefs, null, 2);
+}
+function importPrefs(newPrefsString) {
+    var newPrefs = JSON.parse(newPrefsString);
+    if (typeof newPrefs !== 'object') {
```
Comment (Rob--W) [inline extension/prefs.js]: Let's add `|| newPrefs === null` too (`typeof null` is `object`).

File: extension/options.html [hunk 9]
```diff
@@ -80,6 +86,13 @@ <h1>Open in Browser</h1>
...
+    <input id="importFilePicker" type="file" accept="text/plain">
```
Comment (Rob--W) [inline extension/options.html]: Let's remove the `accept` attribute.

**Comment by Mathnerd314** [comment]
I added error display using window.alert, it seems to work OK. Not sure how to trigger a FileReader error event so I didn't test that part of the code.

**Comment by Mathnerd314** [comment]
Any update on this?

**Comment by Rob--W** [comment]
Sorry for the late reply - I missed the previous notifications. The PR looks good to me.

To test the `fr.onerror`, you could try to select a file for which you don't have the permission to read it. I'm getting a cryptic "Node was not found" error when I do that, but that's better than silently ignoring the error without user feedback.

Thanks for the PR and the follow-up!

**Pull request merged**